aten/src/ATen/core/ATenDispatch.h

```diff
@@ -73,34 +74,108 @@ class CAFFE2_API ATenDispatch {
  public:
   template<class FuncType>
   ATenDispatch& registerOp(Backend backend, const char* schema, FuncType* fn) {
-    std::lock_guard<std::mutex> lock(mutex_);
-    if (op_tables_.find(schema) == op_tables_.end()) {
-      op_tables_.insert(std::make_pair(schema, ATenOpTable(schema)));
+    if (op_was_already_moved_to_c10(schema)) {
+      if (backend == Backend::Undefined) {
+        add_c10_op_registration_(
+          torch::RegisterOperators().op(schema, torch::RegisterOperators::options()
+            .impl_unboxedOnlyCatchAllKernel(fn)
+            .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA)
+        ));
+      } else {
+        add_c10_op_registration_(
+          torch::RegisterOperators().op(schema, torch::RegisterOperators::options()
+            .impl_unboxedOnlyKernel(c10::backendToTensorTypeId(backend), fn)
+            .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA)
+        ));
+      }
+    } else {
+      std::lock_guard<std::mutex> lock(mutex_);
+      if (op_tables_.find(schema) == op_tables_.end()) {
+        op_tables_.insert(std::make_pair(schema, ATenOpTable(schema)));
+      }
+      op_tables_.at(schema).registerOp(backend, reinterpret_cast<void*>(fn));
     }
-    op_tables_.at(schema).registerOp(backend, reinterpret_cast<void*>(fn));
     return *this;
   }
 
   template <class FuncType>
   ATenDispatch& registerVariableOp(const char* schema, FuncType* fn) {
-    std::lock_guard<std::mutex> lock(mutex_);
-    if (op_tables_.find(schema) == op_tables_.end()) {
-      op_tables_.insert(std::make_pair(schema, ATenOpTable(schema)));
+    if (op_was_already_moved_to_c10(schema)) {
+      add_c10_op_registration_(
+        torch::RegisterOperators().op(schema, torch::RegisterOperators::options()
+          .impl_unboxedAutogradKernel(fn)
+          .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA)
+      ));
+    } else {
+      std::lock_guard<std::mutex> lock(mutex_);
+      if (op_tables_.find(schema) == op_tables_.end()) {
+        op_tables_.insert(std::make_pair(schema, ATenOpTable(schema)));
+      }
+      op_tables_.at(schema).registerVariableOp(reinterpret_cast<void*>(fn));
     }
-    op_tables_.at(schema).registerVariableOp(reinterpret_cast<void*>(fn));
     return *this;
   }
 
   const ATenOpTable* getOpTable(const char* schema) const {
+    TORCH_CHECK(!op_was_already_moved_to_c10(schema), "Tried to get the globalAtenDispatch table for an op that is already moved to the c10 dispatcher");;
     auto iter = op_tables_.find(schema);
     TORCH_CHECK(iter != op_tables_.end(),
         "No functions are registered for schema ", schema);
     return &iter->second;
   }
 
  private:
+  static bool op_was_already_moved_to_c10(const char* schema) {
+    auto op_name = parse_operator_name(schema);
+    return aten_ops_already_moved_to_c10().count(op_name) != 0;
+  }
+
+  static c10::OperatorName parse_operator_name(const char* schema) {
```

Comment:
The fact that we have to do this is a big code smell to me. Why do we need `OperatorName` in the first place? The old ATenDispatch.h was OperatorName oblivious; the only thing it knew about was the overall schema string.

Reply:
We need it for checking if a schema is already moved to the c10 dispatcher. This diff has to make ATenDispatch a bit more complicated for that because it has to keep compatibility with backend extensions, but that's ok because we're going to delete ATenDispatch soon.